### PR TITLE
Support for AWS Bedrock & TItan

### PR DESCRIPTION
**Pull request by jogueber**
Thank you for taking time to contribute this pull request!
You might have already read the [contributor guide][1], but as a reminder, please make sure to:

* Sign the [contributor license agreement](https://cla.pivotal.io/sign/spring)
* Rebase your changes on the latest `main` branch and squash your commits
* Add/Update unit tests as needed
* Run a build and make sure all tests pass prior to submission

### CODE REVIEW

**Comment by markpollack** [comment]
Thanks, will review!

File: pom.xml [hunk 1]
```diff
@@ -71,18 +73,18 @@
 		<stringtemplate.version>4.0.2</stringtemplate.version>
 		<open-ai-client.version>0.16.0</open-ai-client.version>
 		<azure-open-ai-client.version>1.0.0-beta.3</azure-open-ai-client.version>
```
Comment (honnuanand) [inline pom.xml:75]: I was told by some folks at AWS that there is a newer release than the 20.157; please make sure to use a more recent, cleaned-up, and better-documented version of the bedrock SDK: https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/bedrock/package-summary.html.

File: spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.experimental.ai.client.AWSBedrockClient;
+
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeClient;
+
+@AutoConfiguration
```
Comment (honnuanand) [inline spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java:13]: I would use the https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/bedrock/BedrockClient.html or the BedrockAsyncClient here. I think just moving into the latest version of the SDK gives you a lot of new classes to use.

**Review by honnuanand — CHANGES_REQUESTED** [review]
I would say , moving into the latest SDK and using all the good functionality there is a good Idea before merging this code.

**Comment by honnuanand** [comment]
@markpollack, please comment on my suggestion of moving to the latest version of the SDK and also using the relevant objects there.

File: spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java (hunk 2 above)
Comment (jogueber) [reply spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java:13]: So just to be sure : the `BedRockClient` provides methods for e.g. deploying models. 
What I currently implements is a way of interacting with the model 
As much as I undersstand `BedRockRuntimeClient` should be correct one to do that right? 
💯  to updating the SDK version

**Comment by jogueber** [comment]
Updated the SDK version. @honnuanand  :)

File: spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java (hunk 2 above)
Comment (honnuanand) [reply spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/aws/AWSBedrockAutoConfiguration.java:13]: You are correct, https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/bedrockruntime/BedrockRuntimeClient.html

thanks

**Comment by jogueber** [comment]
> For some reason, the version on the Pom file still shows.
> 
> <aws-bedrock-client.version>2.20.157</aws-bedrock-client.version>.

sorry my bad should be good now

File: pom.xml (hunk 1 above)
Comment (honnuanand) [reply pom.xml:80]: This is good, thank you!

**Review by honnuanand — APPROVED** [review]

**Comment by markpollack** [comment]
Let me see how I can get an environment up to verify the PR.  Much thanks @jogueber

**Comment by honnuanand** [comment]
I don't have permission to merge. I can try to set up an environment too.
Thanks
Anand


Sent via Superhuman ***@***.***>

**Comment by jogueber** [comment]
Thanks @markpollack  had some time to add more testing and models. 
Ping me if you have any questions!

**Comment by markpollack** [comment]
I'm not sure how to merge this because it is hard to rebase given the `Merge branch 'main' into aws-bedrock-support`.

Is there a way you can get the overall diff between this branch and main to we can follow the git workflow i'm most used to.  I'm open to other suggestions on how to merge this, my git-fu isn't great.  Again, thanks for the follow up, much appreciated.

**Comment by jogueber** [comment]
Sure @markpollack I can also open up a new branch. 
Or we set up a quick Zoom call in the coming days. 
I'm in the European Time Zone and rather flexible :)

**Comment by markpollack** [comment]
Let's arrange a zoom.  Christian would be an easier contact to coordinate.  Please reach out to me via email  mpollack@**ware.com  and ctzolov@**ware.com

**Comment by tzolov** [comment]
Hi @jogueber  
The https://github.com/spring-projects/spring-ai/pull/174 supersedes this PR by providing a more generic coverage of the various Chat and Embedding models provided by Bedrock. 
Once #174  is merged it will bring support for Bedrock: Llama2 Chat, Cohere Chat, Cohere Embedding, Anthropic Chat.
It also provides the implementations of the basic TitanChatBedrockApi, TitanEmbeddingBedrockApi and Ai21Jurassic2ChatBedrockApi API clients that should be used to implement the Ai21 Jurassic2 and Titan's AiClient, AiStreamingClient , EmbeddingClient.

Let me know if you would be interested in implementing the Titan AiClient, AiStreamingClient , EmbeddingClient based on the TitanChatBedrockApi, TitanEmbeddingBedrockApi?

**Comment by tzolov** [comment]
Close as resolved by 88869f10fa54f7ef8b2f0e3135a2acd60477ce70 and 26b1ccdc5768257a0d5feb9243fbee063dd3267f
@jogueber if interested you still can contribute to the incomplete [Bedrock Ai21 Jurassic2 model](https://github.com/spring-projects/spring-ai/blob/main/spring-ai-bedrock/README_JURASSIC2_CHAT.md)

**Pull request closed without merging**